### DISCUSSION TITLE
SQLAlchemy TypeError on hybrid_property condition on datetime

### DISCUSSION THREAD

**Discussion by graille** [sqlalchemy/sqlalchemy#8772] [Usage Questions]
I'm using a SQLAlchemy Hybrid property to determine when a specific event is Finished

```python
@hybrid_property
def is_finished(self):
    """
    Determine if a event can be considered as finished

    A Event is finished if either:
        - "status" is set to "FINISHED"
        - Event has an ending date of more than 30 minutes
    Otherwise, Event is online or interrupted
    :return: State of the event, has a boolean (True = Finished, False = Not finished)
    """

    return (self.status == EventState.FINISHED) \
           | (
                   (self.ending_date != None)
                   & ((datetime.now(tz=pytz.UTC) - timedelta(minutes=30)) >= self.ending_date)
           )
```

However, when called on unfinished Event, I have the following error:

> TypeError: '>=' not supported between instances of 'datetime.datetime'
> and 'NoneType'

My `ending_date` attribute is just a nullable DateTime object

    ending_date: datetime = Column(DateTime, nullable=True, default=None)

I don't understand why this error is raised since the first condition test if the ending_date is not None, so the seocnd condition cannot occurs when it happens.

**Comment by CaselIT** [comment]
Hi,

> I don't understand why this error is raised since the first condition test if the ending_date is not None, so the seocnd condition cannot occurs when it happens.

& does not short circuit, so that part is always evaluated:
```py
import datetime
x = None
(x != None) & (datetime.datetime.now() >= x)
```
raises the same error.

You will probably need to provide a python version (that uses `and` instead of &) and an sql version that's like the above (or could use sqlalcehmy `and_` and `or_`)
